File: 407-trapping-rain-water-ii/trapping-rain-water-ii.java
class Solution {
    public int trapRainWater(int[][] h) {
        int n = h.length;
        int m = h[0].length;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> h[a[0]][a[1]] - h[b[0]][b[1]]);
        boolean[][] vis = new boolean[n][m];

        for(int i = 0 ; i < n ; i ++){
            for(int j = 0 ; j < m ; j ++){
                if(i == 0 || j == 0 || i == n - 1 || j == m - 1){
                    pq.offer(new int[]{i, j});
                }
            }
        }
        int ans = 0 ;

        int[][] di = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        while(!pq.isEmpty()){
            int[] cr = pq.poll() ;
            int x = cr[0] ;
            int y = cr[1] ;
            if(vis[x][y]) continue;
            vis[x][y] = true ;

            for(int i = 0 ; i < 4 ; i ++){
                int nx = x + di[i][0] ;
                int ny = y + di[i][1] ;
                if(nx >= 0 && nx < n && ny >= 0 && ny < m && !vis[nx][ny]){
                    int cur = Math.max(h[x][y] - h[nx][ny], 0) ;
                    h[nx][ny] += cur ;
                    ans += cur ;
                    pq.offer(new int[]{nx, ny}) ;
                }
                
            }
        }
        return ans ;
    }
}

File: 407-trapping-rain-water-ii/trapping-rain-water-ii.py
import heapq
class Solution:
    def trapRainWater(self, h: List[List[int]]) -> int:
        n, m = len(h), len(h[0])
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        q = [] 

        for i in range(n):
            for j in range(m):
                if i == 0 or j == 0 or i == n - 1 or j == m - 1:
                    heapq.heappush(q, (h[i][j], i, j)) # (hight, x, y)
        
        vis = [[False for _ in range(m)] for _ in range(n)]
        ans = 0 
        while q:
            h_, x, y = heapq.heappop(q)
            if vis[x][y]:
                continue
            vis[x][y] = True
            for dx, dy in directions:
                nx, ny = x + dx, y + dy
                if 0 <= nx < n and 0 <= ny < m and not vis[nx][ny]:
                    cr = max(h[x][y] - h[nx][ny], 0)
                    ans += cr 
                    h[nx][ny] += cr 
                    heapq.heappush(q, (h[nx][ny], nx, ny))

        return ans 

